tests/pointer_constraints.cpp

```diff
@@ -0,0 +1,118 @@
+
+using namespace testing;
+using namespace wlcs;
+
+namespace
+{
+auto const any_width = 300;
+auto const any_height = 300;
+auto const nw_middle_x = any_width / 2;
+auto const nw_middle_y = any_height / 2;
+auto const se_middle_x = any_width + nw_middle_x;
+auto const se_middle_y = any_height + nw_middle_y;
+
+struct PointerConstraints : StartedInProcessServer
+{
+    // Create a client with two surface (ne & sw) and a cursor centered on the nw_surface
+    Client client{the_server()};
+    Surface se_surface{client.create_visible_surface(any_width, any_height)};
+    Surface nw_surface{client.create_visible_surface(any_width, any_height)};
+    wl_pointer* const pointer = client.the_pointer();
+
+    Pointer cursor = the_server().create_pointer();
+
+    ZwpPointerConstraintsV1 pointer_constraints{client};
+
+    void SetUp() override
+    {
+        StartedInProcessServer::SetUp();
+
+        // Get the surface in a known position with the cursor over it
+        the_server().move_surface_to(nw_surface, 0, 0);
+        cursor.move_to(nw_middle_x, nw_middle_y);
+        the_server().move_surface_to(se_surface, any_width, any_height);
+    }
+
+    void TearDown() override
+    {
+        client.roundtrip();
+        StartedInProcessServer::TearDown();
+    }
+};
+}
+
+TEST_F(PointerConstraints, can_get_locked_pointer)
+{
+    ZwpLockedPointerV1 locked_ptr{pointer_constraints, nw_surface, pointer, nullptr, ZWP_POINTER_CONSTRAINTS_V1_LIFETIME_ONESHOT};
+    EXPECT_THAT(locked_ptr, NotNull());
+}
+
+TEST_F(PointerConstraints, locked_pointer_on_initially_focussed_surface_gets_locked_notification)
+{
+    ZwpLockedPointerV1 locked_ptr{pointer_constraints, nw_surface, pointer, nullptr, ZWP_POINTER_CONSTRAINTS_V1_LIFETIME_ONESHOT};
+
+    EXPECT_CALL(locked_ptr, locked()).Times(1);
+
+    client.roundtrip();
+}
+
+TEST_F(PointerConstraints, locked_pointer_does_not_move)
+{
+    auto const initial_cursor_position = client.pointer_position();
+    ZwpLockedPointerV1 locked_ptr{pointer_constraints, nw_surface, pointer, nullptr, ZWP_POINTER_CONSTRAINTS_V1_LIFETIME_ONESHOT};
+    EXPECT_CALL(locked_ptr, locked()).Times(AnyNumber());
+    client.roundtrip();
+
+    cursor.move_by(10, 10);
+    client.roundtrip();
+
+    EXPECT_THAT(client.pointer_position(), Eq(initial_cursor_position));
+}
+
+TEST_F(PointerConstraints, locked_pointer_on_initially_unfocussed_surface_gets_no_locked_notification)
+{
+    ZwpLockedPointerV1 locked_ptr{pointer_constraints, se_surface, pointer, nullptr, ZWP_POINTER_CONSTRAINTS_V1_LIFETIME_ONESHOT};
+
+    EXPECT_CALL(locked_ptr, locked()).Times(0);
+
+    client.roundtrip();
+}
+
+TEST_F(PointerConstraints, when_cursor_moves_onto_surface_locked_pointer_gets_locked_notification)
+{
+    ZwpLockedPointerV1 locked_ptr{pointer_constraints, se_surface, pointer, nullptr, ZWP_POINTER_CONSTRAINTS_V1_LIFETIME_ONESHOT};
+
+    EXPECT_CALL(locked_ptr, locked()).Times(1);
+
+    cursor.move_to(se_middle_x, se_middle_y);
```

Comment:
This seems like an implementation detail. In fact, an annoying amount of stuff about how this protocol behaves are implementation-dependent. I'm find with the more basic stuff (locking on a focused surface working immediately), but this is likely to actually be different from compositor to compositor. For example, some might require clicking on the surface before locking.

Reply:
True, the trouble is we don't (yet) have ways to express the requirement to the compositor. I.e. `the_server().capture(cursor, on(se_surface))`.

Reply:
Fair. Maybe add a click here though? I'm not convinced it's desirable for any compositor to capture the pointer on mouse over. It should at least need to be focused, no?

Reply:
Interestingly, adding the click caused my local iteration of the prototype implementation to pass the test.